### PR TITLE
fix(TimePicker-compat): fix freeform behavior + add tests

### PR DESCRIPTION
**Pull request by YuanboXue-Amber**
When typing in a freeform TimePicker and hit enter key, if the input value is not prefix of an option in the dropdown, TimePicker should submit the value as is.
However Combobox behaves differently. It is trying to set the 1st option in listbox as active option (so it will be selected on Enter key). So TimePicker is trying to clear the activeOption state set by combobox.

Before this PR, when typing an invalid option, TimePicker re-renders and re-creates new `children`, triggering [`useEffect` in `useComboboxBaseState`](https://github.com/microsoft/fluentui/blob/61d4a2910a5039af8b8509422ecfe58bc636191c/packages/react-components/react-combobox/src/utils/useComboboxBaseState.ts#L102), setting activeOption to 1st option, and it is reset to undefined by TimePicker.
This PR use set state callback to prevent the unnecessary triggering of `useEffect` in `useComboboxBaseState`. It also adds test for controlled/uncontrolled freeform scenarios.

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/61d4a2910a5039af8b8509422ecfe58bc636191c'>61d4a2910a5039af8b8509422ecfe58bc636191c</a></sub>
<div id="bundleSizeComment9423"/>

**Review by ling1726 — APPROVED** [review]

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluent-ui_VRT_Pipeline_v9"/>

**Comment by YuanboXue-Amber** [comment]
@smhigley please review 🙏 the type change in combobox should be alright [ts demo](https://www.typescriptlang.org/play?ts=4.9.5#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQFA0wCeYWcA8mDMBAHYBqyAGwCuLALxwA3jThwCECAC44qGFGDcA5gG4aAX3pMWABWJh04qTNRYYAQXzAAblnaceACggcu3APxLXH35hLABKJUcIYAATHX04GlweFTgAbRVkGCwAGmUbAGUYTKwAXThxbDwYADoha0LigB5AnmCROAAfOCFuaKwCdSxogD53UMTk+DBTVCUTL3NJazsHZxbuJWWGrN0gA)

File: packages/react-components/react-timepicker-compat-preview/src/components/TimePicker/useTimePicker.tsx [hunk 1]
```diff
@@ -163,17 +163,24 @@ const useStableDateAnchor = (providedDate: Date | undefined, startHour: Hour, en
...
-  const isValueOptionPrefix = value ? options.some(({ text }) => text.indexOf(value) === 0) : false;
   React.useEffect(() => {
-    if (freeform && value && !isValueOptionPrefix) {
-      setActiveOption(undefined);
-    }
-  }, [freeform, isValueOptionPrefix, setActiveOption, value]);
+    setActiveOption(prevActiveOption => {
+      if (!freeform) {
+        return prevActiveOption;
+      }
```
Comment (behowell) [inline packages/react-components/react-timepicker-compat-preview/src/components/TimePicker/useTimePicker.tsx:175]: nit: you can put the `freeform` check outside the call to `setActiveOption`:
```suggestion
    if (freeform) {
      setActiveOption(prevActiveOption => {
```

Comment (behowell) [reply packages/react-components/react-timepicker-compat-preview/src/components/TimePicker/useTimePicker.tsx:175]: You could also move the `value` check outside the call as well:
```ts
    if (freeform && value) {
      setActiveOption(prevActiveOption => {
        if (prevActiveOption?.text?.indexOf(value) === 0)
          return prevActiveOption;
        }
        return undefined;
      });
    }
```

**Review by behowell — APPROVED** [review]

File: packages/react-components/react-timepicker-compat-preview/src/components/TimePicker/useTimePicker.tsx (hunk 1 above)
Comment (YuanboXue-Amber) [reply packages/react-components/react-timepicker-compat-preview/src/components/TimePicker/useTimePicker.tsx]: Yeah this looks cleaner :) updated.

**Pull request merged**